### PR TITLE
[XPU] Enable multiple key kernels for sparse attention

### PR DESCRIPTION
**Pull request by xwu-intel**
## Purpose

This PR migrates sparse indexer helper ops from XPU-local fallbacks to shared custom op wrappers, and aligns the XPU sparse-attention path with vllm-xpu-kernels usage.

Key changes:
- Add shared wrappers in [vllm/_custom_ops.py](vllm/_custom_ops.py) for:
  - `top_k_per_row_prefill`
  - `top_k_per_row_decode`
- Remove Python fallback implementations from [vllm/_xpu_ops.py](vllm/_xpu_ops.py) for:
  - `indexer_k_quant_and_cache`
  - `cp_gather_indexer_k_quant_cache`
  - `top_k_per_row_prefill`
  - `top_k_per_row_decode`
- Update sparse indexer execution flow in [vllm/model_executor/layers/sparse_attn_indexer.py](vllm/model_executor/layers/sparse_attn_indexer.py) to use the unified path and XPU kernels for FP8 MQA logits where applicable.

## Motivation

Reduce duplicated backend-specific Python paths and rely on consolidated custom-op dispatch, while ensuring XPU sparse indexer uses the intended kernels from vllm-xpu-kernels.

## Dependency

- [ ] #39968 
- [x] Release of vllm_xpu_kernels-0.1.8 (https://github.com/vllm-project/vllm-xpu-kernels/releases)

## Test Plan

1. Unit/functional:
- Run sparse indexer and top-k related tests on XPU.
- Run MLA decode/prefill path tests covering:
  - prefill top-k
  - decode top-k
  - fp8 mqa logits
  - fp8 paged mqa logits

2. Suggested commands:
- relevant sparse/MLA indexer tests in your CI matrix on XPU

## Expected Result

- No runtime `AttributeError` from missing `xpu_ops.top_k_per_row_*`.
- XPU sparse-attention path successfully executes FP8 MQA logits / paged FP8 MQA logits without DeepGEMM dependency.

### CODE REVIEW

File: vllm/model_executor/layers/sparse_attn_indexer.py [hunk 1]
```diff
@@ -233,28 +221,16 @@ def sparse_attn_indexer(
...
-                    logits,
-                    next_n,
-                    decode_metadata.seq_lens,
-                    topk_indices,
-                    num_rows,
-                    logits.stride(0),
-                    logits.stride(1),
-                    topk_tokens,
-                )
+            torch.ops._C.top_k_per_row_decode(
```
Comment (jikunshang) [inline vllm/model_executor/layers/sparse_attn_indexer.py]: I am considering we should move this to `_custom_op.py`. same for prefill one.

File: vllm/_xpu_ops.py
Comment (wuxun-zhang) [inline vllm/_xpu_ops.py]: Maybe we also don't need this any more.

Comment (wuxun-zhang) [inline vllm/_xpu_ops.py]: Maybe we don't need this any more.

Comment (wuxun-zhang) [inline vllm/_xpu_ops.py]: same this.

File: vllm/model_executor/layers/sparse_attn_indexer.py (hunk 1 above)
Comment (xwu-intel) [reply vllm/model_executor/layers/sparse_attn_indexer.py]: @wuxun-zhang @jikunshang 
```
if current_platform.is_cuda_alike():
    from vllm import _custom_ops as ops
elif current_platform.is_xpu():
    from vllm._xpu_ops import xpu_ops as ops
```

Maybe we should always import _custom_ops, and keep the fallback ops in xpu_ops. And only call xpu_ops.XXX when the op is not available in vllm-xpu-kernels? If there is a new perf op come up, we can just remove it from xpu_ops and replace.

File: vllm/model_executor/layers/sparse_attn_indexer.py [hunk 2]
```diff
@@ -378,6 +356,29 @@ def forward_cuda(
...
+            weights,
+            self.quant_block_size,
+            self.scale_fmt,
+            self.topk_tokens,
+            self.head_dim,
+            self.max_model_len,
+            self.max_total_seq_len,
+            self.topk_indices_buffer,
+        )
+
```
Comment (yma11) [inline vllm/model_executor/layers/sparse_attn_indexer.py:505]: @chaojun-zhang will vLLM IR can help avoid these op dispatches using duplicated code?

File: vllm/model_executor/layers/sparse_attn_indexer.py [hunk 3]
```diff
@@ -378,6 +356,29 @@ def forward_cuda(
             self.topk_indices_buffer,
         )
 
+    def forward_xpu(
```
Comment (yma11) [inline vllm/model_executor/layers/sparse_attn_indexer.py:497]: So this is exact same call as `forward_cuda` path? I think keeping original is better to avoid heavy duplication.

Comment (xwu-intel) [reply vllm/model_executor/layers/sparse_attn_indexer.py:497]: @yma11 Hi, agree from duplication perspective. but It's my intention to do so to separate the xpu path and not to mix up with cuda path.

Comment (yma11) [reply vllm/model_executor/layers/sparse_attn_indexer.py:497]: Then you can do like this [https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/layers/activation.py#L150](https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/layers/activation.py#L150). This is not good and we have several cases like this. We will see whether vLLM IR can help it.

Comment (xwu-intel) [reply vllm/model_executor/layers/sparse_attn_indexer.py:497]: @yma11 thanks, I will do it for now!

File: vllm/model_executor/layers/sparse_attn_indexer.py (hunk 1 above)
Comment (wuxun-zhang) [reply vllm/model_executor/layers/sparse_attn_indexer.py]: > I am considering we should move this to `_custom_op.py`. same for prefill one.

By doing this, then we can have same path with CUDA, right? No need extra import for XPU platform.

Comment (jikunshang) [reply vllm/model_executor/layers/sparse_attn_indexer.py]: correct. we can share op name with same name. for not same one, we need to figure out how to make it more friendly. it's a valid concern.

Comment (wuxun-zhang) [reply vllm/model_executor/layers/sparse_attn_indexer.py]: I see, make sense.

Comment (xwu-intel) [reply vllm/model_executor/layers/sparse_attn_indexer.py]: top_k_per_row_prefill and top_k_per_row_decode moved to _custom_ops

File: vllm/model_executor/layers/sparse_attn_indexer.py [hunk 4]
```diff
@@ -106,7 +105,7 @@ def sparse_attn_indexer(
         for chunk in prefill_metadata.chunks:
             k_fp8 = k_fp8_full[: chunk.total_seq_lens]
             k_scale = k_scale_full[: chunk.total_seq_lens]
-            ops.cp_gather_indexer_k_quant_cache(
+            xpu_ops.cp_gather_indexer_k_quant_cache(
```
Comment (wuxun-zhang) [inline vllm/model_executor/layers/sparse_attn_indexer.py]: `cp_gather_indexer_k_quant_cache` is also supported in vllm-xpu-kernel 0.1.5. So no need this,

File: vllm/model_executor/layers/sparse_attn_indexer.py (hunk 2 above)
Comment (xinyu-intel) [reply vllm/model_executor/layers/sparse_attn_indexer.py:505]: @ProExpertProg suggestions here? Seems this custom op doesn't have native implementation, how to handle it with vLLM IR?

Comment (yma11) [reply vllm/model_executor/layers/sparse_attn_indexer.py:505]: For this case, it's simple there would be a single `forward` method only and call the registered IR kernel directly regardless platforms. XPU will register its own IR kernel, which should be same one as cuda. Expect no more duplicated code.

File: vllm/_xpu_ops.py [hunk 5]
```diff
@@ -246,118 +246,6 @@ def get_scheduler_metadata(
...
-        k = torch.cat(
-            [k_fp8_bytes, scale_bytes], dim=-1
-        )  # [total_tokens, head_dim + 4]
-
-        slot_mapping = slot_mapping.flatten()
-        # kv_cache: [num_block, block_size, head_dim + 4]
-        kv_cache.view(-1, kv_cache.shape[-1]).index_copy_(0, slot_mapping, k)
-
     @staticmethod
     def cp_gather_indexer_k_quant_cache(
```
Comment (wuxun-zhang) [inline vllm/_xpu_ops.py]: remove this?

File: vllm/model_executor/layers/sparse_attn_indexer.py [hunk 6]
```diff
@@ -307,16 +304,34 @@ def sparse_attn_indexer(
...
-            weights[:num_padded_tokens],
-            seq_lens,
-            decode_metadata.block_table,
-            decode_metadata.schedule_metadata,
-            max_model_len=max_model_len,
-            clean_logits=False,
-        )
+        if current_platform.is_xpu():
+            if not hasattr(torch.ops._xpu_C, "fp8_paged_mqa_logits"):
+                raise RuntimeError("_xpu_C.fp8_paged_mqa_logits is unavailable")
```
Comment (wuxun-zhang) [inline vllm/model_executor/layers/sparse_attn_indexer.py]:
```suggestion
            assert hasattr(torch.ops._xpu_C, "fp8_paged_mqa_logits")
```

File: vllm/model_executor/layers/sparse_attn_indexer.py [hunk 7]
```diff
@@ -220,42 +215,44 @@ def sparse_attn_indexer(
...
-            num_rows = logits.shape[0]
-
-            topk_indices = topk_indices_buffer[
-                chunk.token_start : chunk.token_end, :topk_tokens
-            ]
-
             if current_platform.is_xpu():
-                xpu_ops.top_k_per_row_prefill(  # type: ignore[attr-defined]
-                    logits,
+                if not hasattr(torch.ops._xpu_C, "fp8_mqa_logits"):
```
Comment (jikunshang) [inline vllm/model_executor/layers/sparse_attn_indexer.py]: We should not add this.

File: vllm/model_executor/layers/sparse_attn_indexer.py [hunk 8]
```diff
@@ -220,42 +215,44 @@ def sparse_attn_indexer(
...
+                    raise RuntimeError("_xpu_C.fp8_mqa_logits is unavailable")
+                if q_scale_slice is not None:
+                    raise RuntimeError("XPU fp8_mqa_logits does not support FP4 Q")
+                logits = torch.ops._xpu_C.fp8_mqa_logits(
```
Comment (jikunshang) [inline vllm/model_executor/layers/sparse_attn_indexer.py]: we'd better make it a custom op like what we did in "_xpu_ops.py"

**Comment by wuxun-zhang** [comment]
@xwu-intel can you please fix comments here?

**Review by jikunshang — APPROVED** [review]

**Comment by jikunshang** [comment]
please fix DCO

**Review by wuxun-zhang — APPROVED** [review]
